### PR TITLE
Add test for various length mask functions

### PR DESCRIPTION
**Pull request by rogiervd**
This tests five different functions in SpeechBrain that do essentially the same thing.

## What does this PR do?

Test functions that turn a list of lengths into a mask:
* speechbrain.dataio.dataio.length_to_mask
* speechbrain.lobes.models.transformer.Transformer.get_mask_from_lengths
* speechbrain.nnet.losses.get_mask
* speechbrain.nnet.losses.compute_length_mask
* speechbrain.processing.features.make_padding_mask

Two of the functions fail the test locally (one error is my fault), so assuming this happens in CI as well, I'll add a commit with a fix.

<details>
  <summary><b>Before submitting</b></summary>

- [x] Did you read the [contributor guideline](https://speechbrain.readthedocs.io/en/latest/contributing.html)?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing [tests](https://github.com/speechbrain/speechbrain/tree/develop/tests) pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Does your code adhere to project-specific code style and conventions?

</details>

## PR review

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified
- [ ] Confirm that the changes adhere to compatibility requirements (e.g., Python version, platform)
- [ ] Review the self-review checklist to ensure the code is ready for review

</details>

### CODE REVIEW

**Comment by TParcollet** [comment]
I think we have discussed this merge with @Adel-Moumen and @pplantinga and we agree, they should be merged. I also think that the ones found here are the most used ones.

File: speechbrain/processing/features.py [hunk 1]
```diff
@@ -1545,7 +1545,7 @@ def _load(self, path, end_of_epoch=False):
         self._load_statistics_dict(stats)
 
 
-def make_padding_mask(x, lengths=None, length_dim=1, eps=1e-8):
+def make_padding_mask(x, lengths=None, length_dim=1, eps=1e-5):
```
Comment (TParcollet) [inline speechbrain/processing/features.py:1548]: What is the reason for going from 1e-8 to 1e-5?

**Review by TParcollet — APPROVED** [review]
Thanks Rogier. I honestly don't have much to say, it looks good to me and I don't think it will impact existing models. I'd like a second opinion from @Adel-Moumen however.

File: speechbrain/processing/features.py (hunk 1 above)
Comment (rogiervd) [reply speechbrain/processing/features.py:1548]: `1e-8` is less than 1 unit in the last place in float32. I chose `1e-5` without much thinking.
```
>>> torch.tensor(1., dtype=torch.float32) - 1e-8 - 1
tensor(0.)
>>> torch.tensor(1., dtype=torch.float32) - 1e-7 - 1
tensor(-1.1921e-07)
>>> torch.tensor(1., dtype=torch.float32) - 1e-6 - 1
tensor(-1.0133e-06)
>>> torch.tensor(1., dtype=torch.float32) - 1e-5 - 1
tensor(-1.0014e-05)
```

Should I make it `1e-7` instead you think?

**Review by pplantinga — APPROVED** [review]
Everything LGTM. Thanks for this nice contribution beginning to harmonize the implementations.

**Comment by rogiervd** [comment]
>What is the risk here if eps is too large? Could there be an error if a tensor had a length of 1e5?

> BTW at a sampling rate of 16000 thats about 6 seconds of audio, so we do actually have cases where tensors are that long. But I doubt that it ever matters if we drop the last sample

The test fails at `4e-7` but succeeds at `5e-7` so I'll make it `1e-6`, which gives a better safety margin.

Float32 has 24 bits in the significand. `2**24 == 6e-8` so there could be an argument for doing these fractions at float64.

File: tests/unittests/test_length_mask.py [hunk 2]
```diff
@@ -0,0 +1,598 @@
...
+            low=0, high=max_length + 1, size=lengths_shape
+        )
+        return max_length, lengths
+
+    def generate_lengths():
+        """Generate first adversarial and then random weights."""
+        if adversarial_device is not None:
+            assert fractional_dtype is not None
+            deterministic_lengths = adversarial_fractions(
+                adversarial_device, fractional_dtype, 32
```
Comment (Adel-Moumen) [inline tests/unittests/test_length_mask.py:218]: since you are hardcoding `max_value` to `32` here as well, perhaps leaving it to `32` directly as a default parameter of your function would be better?

File: tests/unittests/test_length_mask.py [hunk 3]
```diff
@@ -375,9 +375,9 @@ def generate_lengths_optional_max_length(
 
 def test_length_to_mask(device):
     """
-    Test speechbrain.dataio.dataio.length_to_mask.
-    length_to_mask takes integer lengths and an optional max_length.
-    It also takes an explicit dtype and device.
+    Test :func:`~speechbrain.lobes.models.transformer.Transformer.get_mask_from_lengths`.
```
[the remark this answers is not in the document]
Comment (Adel-Moumen) [reply tests/unittests/test_length_mask.py:378]: Indeed, good catch! Sorry (I wanted to free you from modifying the docstring by yourself since you did an amazing job already). PS: I am running the pre--commit etc.

**Review by Adel-Moumen — COMMENTED** [review]
LGTM! You did a very nice work @rogiervd. I am quite surprised that we actually didn't had careful checks for length/masking functions. Thanks a lot!

File: tests/unittests/test_length_mask.py [hunk 4]
```diff
@@ -0,0 +1,598 @@
...
+        Torch dtype for which the fractions are adversarial.
+    max_value: int
+        Maximum value (inclusive) of denominator and numerator to be considered.
+
+    Returns
+    -------
+    A dictionary from d to a list of n, where (n/d) * d != n when computed by
+    Torch with the device and dtype passed in, and n < d.
+    """
+    max_value = 32
```
[the remark this answers is not in the document]
Comment (rogiervd) [reply tests/unittests/test_length_mask.py:157]: Whoops! I'll remove this.

**Comment by rogiervd** [comment]
> Indeed, good catch! Sorry (I wanted to free you from modifying the docstring by yourself since you did an amazing job already). PS: I am running the pre--commit etc.

Wouldn't have been a problem. I believe the docstring is now both correct and well-formatted.

File: tests/unittests/test_length_mask.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Adel-Moumen) [reply tests/unittests/test_length_mask.py]: Thanks for the fix @rogiervd and I apologise for the mistake! I corrected it in my for but you were faster!
